File: gateway/resolvers/lastfm.js
const fetch = require('node-fetch')
const {oneLineTrim} = require('common-tags')
const LASTFM_API_KEY = "14cacc2d28210dcd318ffa2085778844"


schema = `
type LastFM_Data {
	name: String
	mbid: String
	duration: Int
	artist: String
	album: String
	artwork: String
	wiki: String
}

extend type Query {
	lastfm(artist:String, title:String): LastFM_Data
}

extend type Track {
	lastfm_data: LastFM_Data
}`


const Track = {
	lastfm_data: fetch_lastfm_data
} 

const Query = {
	lastfm: async (parent, args) => fetch_lastfm_data(args)
}

async function fetch_lastfm_data({artist, title}){
	artist = artist || ''
	title = title || ''

	URL = oneLineTrim`
		http://ws.audioscrobbler.com/2.0/?method=track.getInfo
		&api_key=${LASTFM_API_KEY}
		&artist=${encodeURI(artist.replace(" ", "+"))}
		&track=${encodeURI(title.replace(" ", "+"))}
		&autocorrect
		&format=json`

	request = fetch(URL)
		.then(res => res.json())
		.catch(err => null)
	
	const result = (await request).track

	if(result == null) return {}

	let output = {}
	
	if(result.artist){
		output.artist = result.artist.name	
	}
	
	if(result.album){
		output.album = result.album.name
		output.artwork = result.album.image[2]["#text"]
	}
	
	if(result.wiki){
		output.wiki = result.wiki.summary
	}
	
	output = {...result, ...output}
	return output
}

module.exports = {schema, resolvers:{Query, Track}}

File: src/data_sources/lastfm.rs
use serde::Deserialize;
use surf;

pub async fn lookup_song_name(title: &str, artist: &str) -> Response {
    let url = format!("http://{base}?{method}&api_key={lastfmkey}&artist={artist}&track={track}&autocorrect&format=json",
        base = "ws.audioscrobbler.com/2.0/", 
        lastfmkey = "14cacc2d28210dcd318ffa2085778844",
        method = "method=track.getInfo",
        artist = artist.replace(" ", "+"),
        track = title.replace(" ", "+"),
    );
    
    surf::get(&url).recv_json().await.unwrap()
}

#[derive(Deserialize, Debug)]
pub struct Response {
	pub track: Option<Track>,
}

#[derive(Deserialize, Default, Debug)]
pub struct Track {
	pub name: String,
	pub mbid: Option<String>,
	pub url: Option<String>,
	pub duration: Option<String>,
	pub artist: Option<Artist>,
	pub wiki: Option<Wiki>,
    pub album: Option<Album>,
}

#[derive(Deserialize, Debug, Default)]
pub struct Album {
    pub artist: Option<String>,
    pub title: Option<String>,
    pub mbid: Option<String>,
    pub lastfm_url: Option<String>,
    pub image: Vec<Image>,
}

#[derive(Deserialize, Debug, Default)]
pub struct Image {
    #[serde(rename = "#text")]
    pub url: String,
    pub size: String,
}

#[derive(Deserialize, Clone, Debug, Default)]
pub struct Artist {
	pub name: String,
	pub mbid: Option<String>,
	pub url: Option<String>,
}

#[derive(Deserialize, Clone, Debug, Default)]
pub struct Wiki {
	pub published: Option<String>,
	pub summary: Option<String>,
	pub content: Option<String>,
}

 #[cfg(test)]
 mod tests {
	use super::*;

	#[async_std::test]	
	async fn deserialize_lastfm_response() {
	    let response = lookup_song_name("the distance", "cake").await;	
        println!("{}", response.track.unwrap().album.unwrap().image[2].url);
	}
 }
